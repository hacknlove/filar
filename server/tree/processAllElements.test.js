const { processAllElements } = require("./processAllElements");

const { DOMParser } = require("linkedom");
const parser = new DOMParser();

const { ServerElementsMap } = require("../se/common");
const { createFakeContext } = require("../../test/fakeContext");

describe("processAllElements", () => {
  it("replaces the context in texts", async () => {
    const html = `
<html Title="'My title'">
  <head>
    <title>{{Title}}</title>
  </head>
  <body>
  </body>
</html>
`;
    const document = parser.parseFromString(html);

    await processAllElements(document, createFakeContext());

    expect(document.toString()).toMatchSnapshot();
  });

  it("does not replaces the context in text for SSR", async () => {
    const html = `
  <html Title="'My title'">
    <head>
      <title>{{Title}}</title>
    </head>
    <body>
      <SSR>
        <h1>{{Title}}</h1>
      </SSR>
    </body>
  </html>
  `;
    const document = parser.parseFromString(html);

    await processAllElements(document, createFakeContext());

    expect(document.toString()).toMatchSnapshot();
  });

  it("skips comments", async () => {
    const html = `
  <html Title="'My title'">
    <body>
      <!-- <h1>{{Title}}</h1> -->
    </body>
  </html>
  `;
    const document = parser.parseFromString(html);

    await processAllElements(document, createFakeContext());

    expect(document.toString()).toMatchSnapshot();
  });

  it("replaces text in atrributes", async () => {
    const html = `
      <div AltText="'My title'" ImgSrc="'/img.jpg'">
        <img src={{ImgSrc}} alt={{AltText}}>
      </div>
    `;

    const document = parser.parseFromString(html);

    await processAllElements(document, createFakeContext());

    expect(document.toString()).toMatchSnapshot();
  });

  it("evaluates expressions in text nodes and attributes", async () => {
    const html = `
      <div Title="'My title'">
        <a href="/{{Title.replaceAll(' ', '-')}}">{{Title.toUpperCase()}}</a>
      </div>
    `;

    const document = parser.parseFromString(html);

    await processAllElements(document, createFakeContext());

    expect(document.toString()).toMatchSnapshot();
  });

  it("processes custom elements", async () => {
    const html = `
      <div Title="'My title'">
        <MyComponent Title="'Title'" Content="'Lorem Ipsum'" Href="'/some-page'">
          <h1 slot="title">{{Title}}</h1>
          <p slot="content">{{Content}}</p>
          some text without slot
          <p>some text without slot</p>
        </MyComponent>
      </div>
    `;

    const MyComponent = `
      <div>
        <header>
          <slot name="title"></slot>
        </header>
        <main>
          <slot name="content"></slot>
        </main>
        <aside>
          <slot></slot>
        </aside>
        <footer>
          <a href={{Href}}>Go to page</a>
        </footer>
      </div>
    `;

    const document = parser.parseFromString(html);
    const myComponent = parser.parseFromString(MyComponent).firstElementChild;

    ServerElementsMap.set("MyComponent", myComponent);

    await processAllElements(document, createFakeContext());

    expect(document.toString()).toMatchSnapshot();
  });
});
