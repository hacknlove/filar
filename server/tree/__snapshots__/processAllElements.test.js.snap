// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processAllElements does not replaces the context in text for SSR 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
  <html>
    <head>
      <title>My title</title>
    </head>
    <body>
      <SSR><script class="ssr" type="application/json">{"Title":"My title","Indexes":{"lastIslandId":0,"currentNodeId":0}}</script>
        <h1>{{Title}}</h1>
      </SSR>
    </body>
  </html>
  "
`;

exports[`processAllElements evaluates expressions in text nodes and attributes 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
      <div>
        <a href="/My-title">MY TITLE</a>
      </div>
    "
`;

exports[`processAllElements processes custom elements 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
      <div>
        <div class="MyComponent">
        <header>
          <h1>Title</h1>
        </header>
        <main>
          <p>Lorem Ipsum</p>
        </main>
        <aside>
          
          some text without slot
          <p>some text without slot</p>
        </aside>
        <footer>
          <a href="/some-page">Go to page</a>
        </footer>
      </div>
      </div>
    "
`;

exports[`processAllElements replaces text in atrributes 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
      <div>
        <img src="/img.jpg" alt="My title">
      </img></div>
    "
`;

exports[`processAllElements replaces the context in texts 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<html>
  <head>
    <title>My title</title>
  </head>
  <body>
  </body>
</html>
"
`;

exports[`processAllElements skips comments 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
  <html>
    <body>
      <!-- <h1>{{Title}}</h1> -->
    </body>
  </html>
  "
`;
